sd_bandits/obp_extensions/policy.py

```diff
@@ -85,3 +88,108 @@ def update_params(self, action: int, reward: float) -> None:
         if self.n_trial % self.batch_size == 0:
             self.action_counts = np.copy(self.action_counts_temp)
             self.reward_counts = np.copy(self.reward_counts_temp)
+
+
+@dataclass
+class SegmentPolicy():
+    """Segment Policy.
+    Takes any context-free policy and turns it into a segmented policy, given
+    the number of different segments in the dataset
+
+    Parameters
+    ----------
+    base_policy: BaseContextFreePolicy
+        A policy object that is a subclass of BaseContextFreePolicy
+        Right now, one of Random, EpsilonGreedy, BernoulliTS, or 
+        ExploreThenCommit
+        
+    n_segments: int
+        Number of segments, assumes segments are valued 0 to n_segments-1
+
+    policy_name: str, default=f"{self.policy.policy_name}-seg"
+        Name of bandit policy.
+        
+    """
+    
+    base_policy: BaseContextFreePolicy
+    n_segments: int
+    policy_name: str = ""
+    
+    def __post_init__(self) -> None:
+        """Initialize Class."""
+        assert isinstance(
+            self.base_policy, BaseContextFreePolicy
+        ), "Supplied policy must be an instance of BaseContextFreePolicy"
+        assert isinstance(self.n_segments, int), f"n_segments must be an integer, you supplied a {type(self.n_segments)}"
+        
+        # then instantiate all of base policies attributes
+        self.n_actions = self.base_policy.n_actions
+        self.len_list = self.base_policy.len_list
+        self.batch_size = self.base_policy.batch_size
+        self.random_state = self.base_policy.random_state
+        self.policy_type = "segmented"
+        self.n_trial = 0
+        
+        # give name if no name given
+        if self.policy_name == "":
+            self.policy_name = f'{self.base_policy.policy_name}_seg'
+        
+        # create dictionary of policy objects
+        segment_policy_pairs = [(i, deepcopy(self.base_policy)) for i in range(self.n_segments)]
+        self.segment_policies = dict(segment_policy_pairs)
+        
+    @property
+    def action_counts(self):
+        """Returns action counts"""
+        action_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
+            action_counts += self.segment_policies[i].action_counts
+        return action_counts
+    
+    @property
+    def reward_counts(self):
+        """Returns reward counts"""
+        reward_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
```

Comment:
Can you clarify what this part does? Why do you have to multiply each segment's reward counts by action counts, and why do you divide by all action counts at the end?

Reply:
the 'reward_counts' object isn't actually a count, it's the current estimated reward for each action, which is the total reward for each action divided by the number of times that action has been chosen. So to get that for all estimators, we have to multiply, sum, then divide so everything is on the right scale.
[an example from OBP](https://github.com/st-tech/zr-obp/blob/master/obp/policy/contextfree.py#L93)

Reply:
Ah ok yeah I've seen that bit of code you linked to. (They probably should've named it average rewards or something.) So here you're basically getting the weighted average of each action's observed rewards (across all segments)?

Reply:
Yeah exactly